auctions/english-auction/src/main.sw

```diff
@@ -0,0 +1,377 @@
+
+        // Log the start of the new auction
+        log(AuctionStartEvent {
+            auction: auction, auction_id: storage.total_auctions
+        });
+
+        // Return the auction ID and increment the total auctions counter
+        storage.total_auctions = storage.total_auctions + 1;
+        storage.total_auctions - 1
+    }
+
+    // Uncomment when https://github.com/FuelLabs/fuels-rs/issues/420 is resolved
+    /// Returns the balance of the user's `buy_asset` deposits. If the user has not deposited any
+    /// assets for the provided `auction_id` then `None` will be returned.
+    ///
+    /// # Arguments
+    ///
+    /// * `identity` - The `Identity` of the user which has deposited assets
+    /// * `auction_id` - The `u64` id number of the auction.
+    // #[storage(read)]
+    // fn deposits(identity: Identity, auction_id: u64) -> Option<Asset> {
+    //     storage.deposits.get((identity, auction_id))
+    // }
+
+    /// Allows the users to withdraw their assets if the auction has gone over time, been bought
+    /// outright, or been canceled.
+    ///
+    /// # Arguments
+    ///
+    /// * `auction_id` - The `u64` id number of the auction.
+    ///
+    /// # Reverts
+    ///
+    /// * When the `auction_id` provided does not map to an existing auction.
+    /// * When the duration of the auction has not ended.
+    /// * When the auction's `state` is still in the open bidding state.
+    /// * When the `sender` has already withdrawn their deposit.
+    #[storage(read, write)]fn withdraw(auction_id: u64) {
+        // Make sure this auction exists
+        let auction: Option<Auction> = storage.auctions.get(auction_id);
+        require(auction.is_some(), AccessError::AuctionDoesNotExist);
+        let mut auction = auction.unwrap();
+
+        // Cannot withdraw if the auction is over
+        require(auction.state == State::Closed || height() >= auction.end_block, AccessError::AuctionIsNotClosed);
+
+        // If time has run out set the contract state to closed
+        if (height() >= auction.end_block && auction.state == State::Open) {
+            auction.state = State::Closed;
+            storage.auctions.insert(auction_id, Option::Some(auction));
+        }
+
+        // Set some variables we will need
+        let sender = sender_identity();
+        let bidder: Option<Identity> = auction.highest_bidder;
+        let sender_deposit: Option<Asset> = storage.deposits.get((sender, auction_id));
+
+        // Make sure the sender still has something to withdraw
+        require(sender_deposit.is_some(), UserError::UserHasAlreadyWithdrawn);
+        storage.deposits.insert((sender, auction_id), Option::None());
+        let mut withdrawnAsset = sender_deposit.unwrap();
+
+        // Go ahead and withdraw
+        if (bidder.is_some() && sender == bidder.unwrap()) {
+            // The buyer is withdrawing
+            match auction.sell_asset {
```

Comment:
These will need to be put in an external utils file. A lot of repetitive code here.

Reply:
I saw the control flow and thought "not today... I'll deal with that tomorrow"

Reply:
Yeah it was temporary lol. Already updated